src/counterfactuals/encodings.jl

```diff
@@ -93,11 +104,110 @@ end
 Helper function to decode an array `x` using a data transform `dt::GenerativeModels.AbstractGenerativeModel`.
 """
 function decode_array(
-    data::CounterfactualData, dt::GenerativeModels.AbstractGenerativeModel, x::AbstractArray
+    data::CounterfactualData,
+    dt::GenerativeModels.AbstractGenerativeModel,
+    x::AbstractArray,
 )
     return GenerativeModels.decode(dt, x)
 end
 
+"""
+    run_causal_effects(
+        scm::CausalInference.SCM,
+        x::AbstractArray,
+        idxs::AbstractArray
+    )
+
+Apply the causal effects defined in a structural causal model (SCM) to an array `x`.
+"""
+# function run_causal_effects(scm::CausalInference.SCM, x::AbstractArray, idxs::AbstractArray)
+
+#     g = scm.dag
+
+#     for node in idxs
+#         if node > length(scm.coefficients)
+#             continue
+#         end
+        
+#         coef = scm.coefficients[node]
+#         preds = inneighbors(g, node)
+        
+#         if length(preds) > 0
+#             X = vcat(x[preds, :], ones(1, size(x, 2)))
+
+#             if size(coef, 1) == size(X, 1)
+#                 x[node, :] = coef' * X  # Matrix multiplication
```

Comment:
@JorgeLuizFranco this is where "mutating arrays" happens. 

To solve this, should it not be possible to also get rid of the outer loop over nodes? Then we'd just end up with one big matrix multiplication $X \beta$, where $X$ is the $(1 \times [D+1])$ counterfactual and $([D+1] \times D)$ with $D$ corresponding to the number of features/variables in the SCM. $X$ is $(1 \times [D+1])$ because we append $1$ to account for the bias. The coeff matrix $\beta$ is $([D+1] \times D)$ because for each feature/variable in SCM we get coefficients for each feature plus the bias. 

I might be missing something